linear_operator/operators/_linear_operator.py

```diff
@@ -558,12 +558,7 @@ def _mul_matrix(self, other: Union[torch.Tensor, "LinearOperator"]) -> LinearOpe
         if isinstance(self, DenseLinearOperator) or isinstance(other, DenseLinearOperator):
```

Comment:
This line was the source of the performance regressions. The reasoning behind it appears to be because `MulLinearOperator` *always* computes a root decomposition, which is both inefficient and introduces dead code in its implementation (see below). I am sidestepping this by replacing the `*` with secondary `@` operators in the new special cases of the `DiagLinearOperator` and `BlockDiagLinearOperator` `matmul` methods, leading to `MatmulLinearOperators` instead.

However, this does not get rid of the more general issue. To fix that, I propose two steps in a future PR:
1) Introducing logic in the constructor of `MulLinearOperator` that decides whether or not to build a root decomposition.
2) Even if a root decomposition seems beneficial, delaying its computation until the very last moment when it is needed in `matmul` and cache the result. This will give us ~0 overhead in the case where the linear operator represents a posterior covariance matrix that is constructed via a `posterior` call but only the posterior mean is needed, as was the case in the notebook that exhibited the regression.